### PR TITLE
Avoid overriding model_type in TasksManager

### PR DESCRIPTION
**Pull request by fxmarty**
This is bad practice as `PretrainedConfig.model_type` is a class attribute.

We also dissociate `_SUPPORTED_MODEL_TYPE` by library, as suggested by @JingyaHuang.

Both
```
optimum-cli export onnx -m sentence-transformers/all-MiniLM-L6-v2 minilm_st_onnx --library-name sentence_transformers
```
and
```
optimum-cli export onnx -m sentence-transformers/all-MiniLM-L6-v2 minilm_onnx --library-name transformers
```

work as expected.

Still need to specify `transformers` in the `ORTModel` with `export=True` (we do not support sentence-transformers).

### CODE REVIEW

**Review by JingyaHuang — COMMENTED** [review]
May I start testing with sub package like neuron?

File: optimum/exporters/tasks.py [hunk 1]
```diff
@@ -292,6 +292,46 @@ class TasksManager:
...
+        "default-timm-config": supported_tasks_mapping("image-classification", onnx="TimmDefaultOnnxConfig"),
+    }
+
+    _SENTENCE_TRANSFORMERS_SUPPORTED_MODEL_TYPE = {
+        "clip": supported_tasks_mapping(
+            "feature-extraction",
+            "sentence-similarity",
+            onnx="SentenceTransformersCLIPOnnxConfig",
+        ),
+        "transformer": supported_tasks_mapping(
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/exporters/tasks.py:328]: @JingyaHuang We don't override `model_type`, but simply add an attribute `export_model_type` to avoid to have multiple duplicate keys in this dictionary (or in timm dictionary).

Anyway, having `AutoConfig` in `modeling_base.py` is probably not a good thing.

File: tests/exporters/exporters_utils.py [hunk 2]
```diff
@@ -322,6 +322,6 @@
 }
 
 PYTORCH_SENTENCE_TRANSFORMERS_MODEL = {
-    "sentence-transformers-clip": "sentence-transformers/all-MiniLM-L6-v2",
-    "sentence-transformers-transformer": "sentence-transformers/clip-ViT-B-32-multilingual-v1",
+    "clip": "sentence-transformers/clip-ViT-B-32",
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply tests/exporters/exporters_utils.py:325]: Thank you, indeed would fail for local checkpoint, fixed

**Comment by fxmarty** [comment]
@JingyaHuang Sure, feel free to test so that this PR does not break too many things. What I expect to break is the usage of `TasksManager.create_register` in case you want to register models for export for libraries other than transformers (that is default): https://github.com/huggingface/optimum-neuron/blob/f81c3654c358eb2acfc1d294f4e43725ea822044/optimum/exporters/neuron/model_configs.py#L69

e.g. you would need @register_in_tasks_manager("whatever", *DIFFUSERS_TASKS, library_name="diffusers")

**Comment by JingyaHuang** [comment]
AFAIK, so far the register is only used under neuron. @fxmarty

File: optimum/exporters/onnx/utils.py [hunk 3]
```diff
@@ -323,6 +323,7 @@ def get_stable_diffusion_models_for_export(
         text_encoder_config_constructor = TasksManager.get_exporter_config_constructor(
             model=pipeline.text_encoder,
             exporter="onnx",
+            library_name="diffusers",
```
Comment (JingyaHuang) [inline optimum/exporters/onnx/utils.py:326]: the `library_name` of the clip text model should be transformers: https://huggingface.co/stabilityai/stable-diffusion-2-1/blob/main/model_index.json. In the pipeline, only unet and vae's implementation classes are from diffusers.

File: optimum/exporters/onnx/utils.py [hunk 4]
```diff
@@ -366,6 +370,7 @@ def get_stable_diffusion_models_for_export(
         onnx_config_constructor = TasksManager.get_exporter_config_constructor(
             model=pipeline.text_encoder_2,
             exporter="onnx",
+            library_name="diffusers",
```
Comment (JingyaHuang) [inline optimum/exporters/onnx/utils.py:373]: same for text encoder 2: https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/model_index.json

**Review by JingyaHuang — COMMENTED** [review]
Thanks for the PR, left some questions and nits. Testing for neuron as well, will update you in a bit.

File: optimum/exporters/onnx/__main__.py [hunk 5]
```diff
@@ -398,7 +402,13 @@ def main_export(
                     "Could not infer the pad token id, which is needed in this case, please provide it with the --pad_token_id argument"
                 )
 
-    model_type = "stable-diffusion" if "stable-diffusion" in task else model.config.model_type.replace("_", "-")
+    if "stable-diffusion" in task:
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/exporters/onnx/__main__.py:405]: cc @echarlaix

File: optimum/exporters/onnx/__main__.py [hunk 6]
```diff
@@ -76,9 +76,8 @@ def _get_submodels_and_onnx_configs(
     library_name: str = "transformers",
     model_kwargs: Optional[Dict] = None,
 ):
-    is_stable_diffusion = "stable-diffusion" in task
     if not custom_architecture:
-        if is_stable_diffusion:
+        if library_name == "diffusers":
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/exporters/onnx/__main__.py:80]: @JingyaHuang Made `library_name` non-optional.

File: optimum/exporters/onnx/utils.py (hunk 3 above)
Comment (fxmarty) [reply optimum/exporters/onnx/utils.py:326]: @JingyaHuang `clip-text-model` can not be exported through `main_export` (see https://github.com/huggingface/optimum/blob/45c1c095e6bbf498ea2e10d3d8449858e037b76c/optimum/exporters/tasks.py#L1056), and is currently diffusers-specific: https://github.com/fxmarty/optimum/blob/22286de80b1da7135566bea5fafa2c04145b8961/optimum/exporters/tasks.py#L296-L303

File: optimum/exporters/onnx/utils.py (hunk 4 above)
Comment (fxmarty) [reply optimum/exporters/onnx/utils.py:373]: same

File: optimum/exporters/tasks.py [hunk 7]
```diff
@@ -1125,42 +1151,58 @@ def get_supported_tasks_for_model_type(
...
+            library_name (`Optional[str]`, defaults to `None`):
+                The library name of the model. Can be any of "transformers", "timm", "diffusers", "sentence_transformers".
 
         Returns:
             `TaskNameToExportConfigDict`: The dictionary mapping each task to a corresponding `ExportConfig`
             constructor.
         """
+        if library_name is None:
+            logger.warning(
+                'Not passing the argument `library_name` to `get_supported_tasks_for_model_type` is deprecated and the support will be removed in a future version of Optimum. Please specify a `library_name`. Defaulting to `"transformers`.'
```
Comment (JingyaHuang) [inline optimum/exporters/tasks.py:1163]: I feel the warning a little bit misleading, feeling like that we are about to deprecate the func which is not the case. Maybe just rephrase something some "`library_name` will be mandatory"?

File: optimum/exporters/onnx/utils.py (hunk 3 above)
Comment (fxmarty) [reply optimum/exporters/onnx/utils.py:326]: @JingyaHuang `_UNSUPPORTED_CLI_MODEL_TYPE` was introduced by @echarlaix for the stable diffusion export, where some entries (unet, etc.) were added in `_SUPPORTED_MODEL_TYPES` but could nevertheless not be exported with `main_export`.

It is not particularly confusing, given that as I linked (https://github.com/fxmarty/optimum/blob/22286de80b1da7135566bea5fafa2c04145b8961/optimum/exporters/tasks.py#L296-L303) the `clip-text-model` is specific to the export of diffusers models, just like it was before.

This [does not matter](https://github.com/huggingface/optimum/blob/da6f9e2d9bc57c4a337ac3a30a831f295325a199/optimum/exporters/onnx/__main__.py#L390) for the model loading since `get_model_from_task` `get_model_class_for_task` just use the top level `library_name` to load the full pipeline. This is just a matter of loading the ONNX config.

If we want to change that, I suggest to do so in an other PR.

File: optimum/exporters/tasks.py (hunk 7 above)
Comment (fxmarty) [reply optimum/exporters/tasks.py:1163]: Thank you, fixed in https://github.com/huggingface/optimum/pull/1647/commits/665c880d547a3664fe60fd8ca8d931d5b68b1b73

**Review by JingyaHuang — DISMISSED** [review]
LGTM, thanks for the PR!

File: optimum/exporters/tasks.py [hunk 8]
```diff
@@ -1624,10 +1668,13 @@ def infer_library_from_model(
                     library_name = "timm"
                 elif hasattr(model_config, "_diffusers_version"):
                     library_name = "diffusers"
-                elif any(file_path.startswith("sentence_") for file_path in all_files):
-                    library_name = "sentence_transformers"
-                else:
```
Comment (JingyaHuang) [inline optimum/exporters/tasks.py:1629]: can we add back the case of transformers? When testing the inference, it seems that library could be None in this case whereas it should be transformers:
```
============================================================ FAILURES ============================================================
__________________ NeuronModelForFeatureExtractionIntegrationTest.test_compare_to_transformers_dyn_bs_0_albert ___________________

a = (<tests.inference.test_modeling.NeuronModelForFeatureExtractionIntegrationTest testMethod=test_compare_to_transformers_dyn_bs_0_albert>,)
kw = {}

    @wraps(func)
    def standalone_func(*a, **kw):
>       return func(*(a + p.args), **p.kwargs, **kw)

../pyvenv/aws_neuron_venv_2.16/lib/python3.8/site-packages/parameterized/parameterized.py:620: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/inference/test_modeling.py:197: in test_compare_to_transformers_dyn_bs
    neuron_model_dyn = NeuronModelForFeatureExtraction.from_pretrained(
../pyvenv/aws_neuron_venv_2.16/lib/python3.8/site-packages/optimum/modeling_base.py:349: in from_pretrained
    library_name = TasksManager.infer_library_from_model(model_id, subfolder, revision, cache_dir)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'optimum.exporters.tasks.TasksManager'>, model_name_or_path = '/tmp/albert_dyn_bs_true_feature-extraction_r02o_ps_'
subfolder = '', revision = None, cache_dir = None, library_name = None

    @classmethod
    def infer_library_from_model(
        cls,
        model_name_or_path: Union[str, Path],
        subfolder: str = "",
        revision: Optional[str] = None,
        cache_dir: str = huggingface_hub.constants.HUGGINGFACE_HUB_CACHE,
        library_name: Optional[str] = None,
    ):
        """
        Infers the library from the model repo.
    
        Args:
            model_name_or_path (`str`):
                The model to infer the task from. This can either be the name of a repo on the HuggingFace Hub, an
                instance of a model, or a model class.
            subfolder (`str`, defaults to `""`):
                In case the model files are located inside a subfolder of the model directory / repo on the Hugging
                Face Hub, you can specify the subfolder name here.
            revision (`Optional[str]`, *optional*, defaults to `None`):
                Revision is the specific model version to use. It can be a branch name, a tag name, or a commit id.
            cache_dir (`Optional[str]`, *optional*):
                Path to a directory in which a downloaded pretrained model weights have been cached if the standard cache should not be used.
            library_name (`Optional[str]`, *optional*):
                The library name of the model. Can be any of "transformers", "timm", "diffusers", "sentence_transformers".
        Returns:
            `str`: The library name automatically detected from the model repo.
        """
        if library_name is not None:
            return library_name
    
        full_model_path = Path(model_name_or_path) / subfolder
    
        if not full_model_path.is_dir():
            model_info = huggingface_hub.model_info(model_name_or_path, revision=revision)
            library_name = getattr(model_info, "library_name", None)
    
            # sentence-transformers package name is sentence_transformers
            if library_name is not None:
                library_name = library_name.replace("-", "_")
    
        if library_name is None:
            all_files, _ = TasksManager.get_model_files(model_name_or_path, subfolder, cache_dir)
    
            if "model_index.json" in all_files:
                library_name = "diffusers"
            elif CONFIG_NAME in all_files:
                model_config = PretrainedConfig.from_pretrained(
                    model_name_or_path, subfolder=subfolder, revision=revision
                )
    
                if hasattr(model_config, "pretrained_cfg") or hasattr(model_config, "architecture"):
                    library_name = "timm"
                elif hasattr(model_config, "_diffusers_version"):
                    library_name = "diffusers"
            elif (
                any(file_path.startswith("sentence_") for file_path in all_files)
                or "config_sentence_transformers.json" in all_files
            ):
                library_name = "sentence_transformers"
            else:
                library_name = "transformers"
    
        if library_name is None:
>           raise ValueError(
                "The library name could not be automatically inferred. If using the command-line, please provide the argument --library {transformers,diffusers,timm,sentence_transformers}. Example: `--library diffusers`."
            )
E           ValueError: The library name could not be automatically inferred. If using the command-line, please provide the argument --library {transformers,diffusers,timm,sentence_transformers}. Example: `--library diffusers`.

../pyvenv/aws_neuron_venv_2.16/lib/python3.8/site-packages/optimum/exporters/tasks.py:1680: ValueError
```

Comment (fxmarty) [reply optimum/exporters/tasks.py:1629]: For which model is that? I don't see how this could occur.

Comment (JingyaHuang) [reply optimum/exporters/tasks.py:1629]: It's `hf-internal-testing/tiny-random-AlbertModel` and all inference tests of encoder models fail.

```
(Pdb) all_files
['special_tokens_map.json', 'tokenizer.json', 'config.json', 'tokenizer_config.json', 'spiece.model', 'model.neuron']
(Pdb) CONFIG_NAME in all_files
True
(Pdb) model_config
PretrainedConfig {
  "_name_or_path": "hf-internal-testing/tiny-random-AlbertModel",
  "architectures": [
    "AlbertModel"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": 2,
  "classifier_dropout_prob": 0.1,
  "embedding_size": 128,
  "eos_token_id": 3,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 36,
  "initializer_range": 0.02,
  "inner_group_num": 1,
  "intermediate_size": 37,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "neuron": {
    "auto_cast": null,
    "auto_cast_type": null,
    "compiler_type": "neuronx-cc",
    "compiler_version": "2.12.54.0+f631c2365",
    "disable_fallback": false,
    "disable_fast_relayout": false,
    "dynamic_batch_size": true,
    "input_names": [
      "input_ids",
      "attention_mask",
      "token_type_ids"
    ],
    "model_type": "albert",
    "optlevel": "2",
    "output_attentions": false,
    "output_hidden_states": false,
    "output_names": [
      "last_hidden_state",
      "pooler_output"
    ],
    "static_batch_size": 1,
    "static_sequence_length": 32
  },
  "num_attention_heads": 6,
  "num_hidden_groups": 6,
  "num_hidden_layers": 6,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "task": "feature-extraction",
  "torch_dtype": "float32",
  "transformers_version": "4.36.2",
  "type_vocab_size": 16,
  "vocab_size": 30000
}
```

Comment (JingyaHuang) [reply optimum/exporters/tasks.py:1629]: Well maybe it's for neuron, not sure how onnx files are organised. Maybe we should enable passing lib name to from_pretrained as well?

Comment (fxmarty) [reply optimum/exporters/tasks.py:1629]:
```python
from optimum.exporters import TasksManager

library = TasksManager.infer_library_from_model("hf-internal-testing/tiny-random-AlbertModel")
print(library)
```

works as expected. Could you provide a reproduction?

Comment (fxmarty) [reply optimum/exporters/tasks.py:1629]: Thank you @JingyaHuang, this should be fixed, could you give a try?

Comment (JingyaHuang) [reply optimum/exporters/tasks.py:1629]: Yay it works! Thxxxxx!

**Comment by JingyaHuang** [comment]
Any update on this? 👀

**Comment by fxmarty** [comment]
Merging as failing tests are unrelated & fixed in https://github.com/huggingface/optimum/pull/1683 & https://github.com/microsoft/onnxruntime/pull/19421
